From 6ef136ce3b1fa9537df2d2040e436f59c0f9436f Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Tue, 20 Mar 2018 20:10:15 +0100
Subject: [PATCH] MXC-IPU: Fix buffer addressing for ENGR00161315

Fix the source address calculation for down stripes and invalidate
temporary buffers after re-allocation.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/mxc/ipu3/ipu_device.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/ipu3/ipu_device.c b/drivers/mxc/ipu3/ipu_device.c
index eb1c5f7..6f3fad6 100644
--- a/drivers/mxc/ipu3/ipu_device.c
+++ b/drivers/mxc/ipu3/ipu_device.c
@@ -2463,6 +2463,7 @@ static void vdi_split_process(struct ipu_soc *ipu, struct ipu_task_entry *t)
 			return;
 		}
 		memset(parent->vditmpbuf[0], 0, vdi_size);
+		parent->buf0filled = false;
 
 		parent->vditmpbuf[1] = kmalloc(vdi_size, GFP_KERNEL);
 		if (parent->vditmpbuf[1] == NULL) {
@@ -2472,6 +2473,7 @@ static void vdi_split_process(struct ipu_soc *ipu, struct ipu_task_entry *t)
 			return;
 		}
 		memset(parent->vditmpbuf[1], 0, vdi_size);
+		parent->buf1filled = false;
 
 		parent->old_save_lines = vdi_save_lines;
 		parent->old_size = vdi_size;
@@ -2525,7 +2527,7 @@ static void vdi_split_process(struct ipu_soc *ipu, struct ipu_task_entry *t)
 	else if ((stripe_mode == DOWN_STRIPE) ||
 			(stripe_mode == (DOWN_STRIPE | LEFT_STRIPE))) {
 		if (!parent->buf0filled) {
-			offset_addr = t->set.o_off + vdi_save_lines*t->set.ostride;
+			offset_addr = t->set.o_off - vdi_save_lines*t->set.ostride;
 			dmac_flush_range(base_off + offset_addr,
 					base_off + offset_addr + vdi_size);
 			outer_flush_range(t->output.paddr + offset_addr,
@@ -2583,7 +2585,7 @@ static void vdi_split_process(struct ipu_soc *ipu, struct ipu_task_entry *t)
 	/*Down stripe or Down&Right stript*/
 	else if (stripe_mode == (DOWN_STRIPE | RIGHT_STRIPE)) {
 		if (!parent->buf1filled) {
-			offset_addr = t->set.o_off + vdi_save_lines*t->set.ostride;
+			offset_addr = t->set.o_off - vdi_save_lines*t->set.ostride;
 			dmac_flush_range(base_off + offset_addr,
 					base_off + offset_addr + vdi_save_lines*t->set.ostride);
 			outer_flush_range(t->output.paddr + offset_addr,
-- 
2.7.4

